File: implementations/javascript/src/StateSyncLayer.js
// State Synchronization Layer Implementation
// This module implements the state synchronization layer using CRDTs
// and integrates with the identity layer for node identification

import IdentityLayer from './index.js';
import { CRDT, LWWRegister, ORSet } from './state-sync/index.js';

class StateSyncLayer {
  constructor() {
    // Initialize identity layer
    this.identityLayer = new IdentityLayer();
    
    // Generate Nostr keys for this node
    this.identityLayer.generateNostrKeys();
    
    // Create DID document for this node
    this.didDocument = this.identityLayer.createDIDDocument();
    
    // Node ID is the DID
    this.nodeId = this.identityLayer.did;
    
    // Store for CRDTs
    this.crdts = new Map();
  }

  // Create a new CRDT
  createCRDT(id, type, initialValue = null) {
    let crdt;
    
    switch (type) {
      case 'LWWRegister':
        crdt = new LWWRegister(id, this.nodeId, initialValue);
        break;
      case 'ORSet':
        crdt = new ORSet(id, this.nodeId);
        // If initial value is provided for ORSet, add it
        if (initialValue !== null && Array.isArray(initialValue)) {
          initialValue.forEach(element => {
            crdt.add(element, this.nodeId);
          });
        }
        break;
      default:
        throw new Error(`Unsupported CRDT type: ${type}`);
    }
    
    this.crdts.set(id, crdt);
    return crdt;
  }

  // Get a CRDT by ID
  getCRDT(id) {
    return this.crdts.get(id);
  }

  // Update a CRDT
  updateCRDT(id, operation) {
    const crdt = this.crdts.get(id);
    if (!crdt) {
      throw new Error(`CRDT with id ${id} not found`);
    }
    
    // Apply the operation to the CRDT
    crdt.applyOperation(operation);
    
    // Return the updated CRDT
    return crdt;
  }

  // Merge with another CRDT
  mergeCRDT(id, otherCRDT) {
    const crdt = this.crdts.get(id);
    if (!crdt) {
      // If the CRDT doesn't exist, create it
      this.crdts.set(id, otherCRDT);
      return otherCRDT;
    }
    
    // Merge with the existing CRDT
    crdt.merge(otherCRDT);
    return crdt;
  }

  // Serialize all CRDTs to JSON
  serializeState() {
    const state = {};
    for (const [id, crdt] of this.crdts) {
      state[id] = crdt.toJSON();
    }
    return state;
  }

  // Deserialize state from JSON
  deserializeState(state) {
    for (const [id, json] of Object.entries(state)) {
      let crdt;
      
      // Create the appropriate CRDT based on type
      switch (json.type) {
        case 'LWWRegister':
          crdt = LWWRegister.fromJSON(json);
          break;
        case 'ORSet':
          crdt = ORSet.fromJSON(json);
          break;
        default:
          throw new Error(`Unsupported CRDT type: ${json.type}`);
      }
      
      this.crdts.set(id, crdt);
    }
  }

  // Get the node ID (DID)
  getNodeId() {
    return this.nodeId;
  }

  // Get the DID document
  getDIDDocument() {
    return this.didDocument;
  }

  // Resolve a DID
  async resolveDID(did) {
    return await this.identityLayer.resolveDID(did);
  }
}

export default StateSyncLayer;

File: implementations/python/src/state_sync_layer.py
"""
State Synchronization Layer Implementation
This module implements the state synchronization layer using CRDTs
and integrates with the identity layer for node identification
"""

import json
from .identity_layer import IdentityLayer
from .state_sync.crdt import CRDT
from .state_sync.lww_register import LWWRegister
from .state_sync.or_set import ORSet


class StateSyncLayer:
    def __init__(self):
        # Initialize identity layer
        self.identity_layer = IdentityLayer()
        
        # Generate Nostr keys for this node
        self.identity_layer.generate_nostr_keys()
        
        # Create DID document for this node
        self.did_document = self.identity_layer.create_did_document()
        
        # Node ID is the DID
        self.node_id = self.identity_layer.did
        
        # Store for CRDTs
        self.crdts = {}
        
    def create_crdt(self, id, type, initial_value=None):
        """Create a new CRDT"""
        crdt = None
        
        if type == 'LWWRegister':
            crdt = LWWRegister(id, self.node_id, initial_value)
        elif type == 'ORSet':
            crdt = ORSet(id, self.node_id)
            # If initial value is provided for ORSet, add it
            if initial_value is not None and isinstance(initial_value, list):
                for element in initial_value:
                    crdt.add(element, self.node_id)
        else:
            raise Exception(f"Unsupported CRDT type: {type}")
            
        self.crdts[id] = crdt
        return crdt
        
    def get_crdt(self, id):
        """Get a CRDT by ID"""
        return self.crdts.get(id)
        
    def update_crdt(self, id, operation):
        """Update a CRDT"""
        crdt = self.crdts.get(id)
        if crdt is None:
            raise Exception(f"CRDT with id {id} not found")
            
        # Apply the operation to the CRDT
        crdt.apply_operation(operation)
        
        # Return the updated CRDT
        return crdt
        
    def merge_crdt(self, id, other_crdt):
        """Merge with another CRDT"""
        crdt = self.crdts.get(id)
        if crdt is None:
            # If the CRDT doesn't exist, create it
            self.crdts[id] = other_crdt
            return other_crdt
            
        # Merge with the existing CRDT
        crdt.merge(other_crdt)
        return crdt
        
    def serialize_state(self):
        """Serialize all CRDTs to JSON"""
        state = {}
        for id, crdt in self.crdts.items():
            state[id] = crdt.to_json()
        return state
        
    def deserialize_state(self, state):
        """Deserialize state from JSON"""
        for id, json_data in state.items():
            crdt = None
            
            # Create the appropriate CRDT based on type
            if json_data['type'] == 'LWWRegister':
                crdt = LWWRegister.from_json(json_data)
            elif json_data['type'] == 'ORSet':
                crdt = ORSet.from_json(json_data)
            else:
                raise Exception(f"Unsupported CRDT type: {json_data['type']}")
                
            self.crdts[id] = crdt
            
    def get_node_id(self):
        """Get the node ID (DID)"""
        return self.node_id
        
    def get_did_document(self):
        """Get the DID document"""
        return self.did_document
        
    def resolve_did(self, did):
        """Resolve a DID"""
        return self.identity_layer.resolve_did(did)